File: packages/pwa/src/lib/ulid.ts
/**
 * Browser-safe ULID generator — Crockford Base32, 26 chars, 48-bit
 * millisecond timestamp + 80 random bits. Matches the scheme used in
 * packages/shared-schema/src/ulid.ts and packages/sync-cli/internal/cloud.
 */

const ENC = '0123456789ABCDEFGHJKMNPQRSTVWXYZ';

export function newUlid(): string {
  const ts = BigInt(Date.now());
  const tsB = ts.toString(2).padStart(48, '0');
  const rnd = crypto.getRandomValues(new Uint8Array(10));
  let rndB = '';
  for (const b of rnd) rndB += b.toString(2).padStart(8, '0');
  const bits = tsB + rndB; // 128
  let out = '';
  for (let i = 0; i < 26; i++) {
    const slice = bits.slice(i * 5, i * 5 + 5).padEnd(5, '0');
    out += ENC[parseInt(slice, 2)];
  }
  return out;
}

File: packages/mac-native/Sources/PalmKit/Ulid.swift
import Foundation

/// Crockford Base32 ULID — 26 chars, 48-bit ms timestamp + 80 random bits.
/// Lexicographically sortable by creation time. Matches the ID scheme used
/// by the PWA (`src/lib/ulid.ts`), the Android app (`data/Ulid.kt`) and the
/// Go sync-cli so that client-generated IDs never collide across devices.
public enum Ulid {
    private static let enc = Array("0123456789ABCDEFGHJKMNPQRSTVWXYZ")

    public static func new(date: Date = Date()) -> String {
        let ms = UInt64((date.timeIntervalSince1970 * 1000).rounded())
        var chars = [Character](repeating: "0", count: 26)

        // 48-bit time → first 10 base32 chars (most-significant first).
        var t = ms
        var i = 9
        while i >= 0 {
            chars[i] = enc[Int(t & 0x1F)]
            t >>= 5
            i -= 1
        }
        // 80 bits of randomness → last 16 chars.
        for j in 10..<26 {
            chars[j] = enc[Int.random(in: 0..<32)]
        }
        return String(chars)
    }
}
